tools/rpmuncompress.c

```diff
@@ -151,26 +155,47 @@ static char *doUntar(const char *fn)
     if ((at = getArchiver(fn)) == NULL)
 	goto exit;
 
-    if (dstpath) {
-	    int sr = singleRoot(fn);
+    int needtar = (at->extractable == 0);
 
-	    /* the trick is simple, if the archive has multiple entries,
-	     * just extract it into the specified destination path, otherwise
-	     * strip the first path entry and extract in the destination path
-	     */
-	    rasprintf(&mkdir, "mkdir '%s' ; ", dstpath);
+    if (dstpath) {
+	char * sr = singleRoot(fn);
+
+	/* the trick is simple, if the archive has multiple entries,
+	 * just extract it into the specified destination path, otherwise
+	 * strip the first path entry and extract in the destination path
+	 */
+	rasprintf(&mkdir, "mkdir '%s' ; ", dstpath);
+	if (needtar)
 	    rasprintf(&stripcd, " -C '%s' %s", dstpath, sr ? "--strip-components=1" : "");
+	else {
+	    char * moveup;
+	    if (sr)
+		/* rename top dir to tmp name to avoid collisions */
+		/* then move files within one level up */
+		rasprintf(
+		    &moveup,
+		    " && "
+		    "tmp=`mktemp -u -p'%s'` && "
+		    "mv %s/%s $tmp && "
```

Comment:
Shouldn't we be quoting all these paths, like the mktemp -p is?

Reply:
Done.